File: golang/confluence_secrets_scanner.go
package main

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"regexp"
)

const (
	confluenceURL = "https://your-confluence-instance.com"
)

var (
	patterns = []string{
		"[a-zA-Z0-9]{32}",
		"(?i)password\\s*[:=]\\s*[a-zA-Z0-9!@#$%^&*()_+-=]{6,}",
		"Bearer\\s+[a-zA-Z0-9\\-_]+\\.[a-zA-Z0-9\\-_]+\\.[a-zA-Z0-9\\-_]+",
	}
)

type Page struct {
	ID     string `json:"id"`
	Title  string `json:"title"`
	Body   Body   `json:"body"`
}

type Body struct {
	Storage Storage `json:"storage"`
}

type Storage struct {
	Value string `json:"value"`
}

type SearchResult struct {
	Results []Page `json:"results"`
	Size    int    `json:"size"`
}

type SpaceResult struct {
	Results []Space `json:"results"`
	Size    int     `json:"size"`
}

type Space struct {
	Key string `json:"key"`
}

func main() {
	accessToken := os.Getenv("CONFLUENCE_ACCESS_TOKEN")
	client := &http.Client{}

	spaces := getAllSpaces(client, accessToken)
	for _, space := range spaces {
		fmt.Printf("Searching for secrets in space \"%s\"...\n", space.Key)
		searchSpace(space.Key, client, accessToken)
	}

	fmt.Println("Finished searching for secrets.")
}

func getAllSpaces(client *http.Client, accessToken string) []Space {
	var spaces []Space
	start := 0
	limit := 25

	for {
		url := fmt.Sprintf("%s/rest/api/space?start=%d&limit=%d", confluenceURL, start, limit)
		req, _ := http.NewRequest("GET", url, nil)
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", accessToken))

		resp, err := client.Do(req)
		if err != nil {
			fmt.Println("Error fetching spaces:", err)
			os.Exit(1)
		}
		defer resp.Body.Close()

		body, _ := ioutil.ReadAll(resp.Body)

		var spaceResult SpaceResult
		json.Unmarshal(body, &spaceResult)
		spaces = append(spaces, spaceResult.Results...)

		if spaceResult.Size < limit {
			break
		}

		start += limit
	}

	return spaces
}

func searchSpace(spaceKey string, client *http.Client, accessToken string) {
	start := 0
	limit := 25

	for {
		url := fmt.Sprintf("%s/rest/api/content/search?cql=space=%s&start=%d&limit=%d&expand=id", confluenceURL, spaceKey, start, limit)
		req, _ := http.NewRequest("GET", url, nil)
		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", accessToken))

		resp, err := client.Do(req)
		if err != nil {
			fmt.Println("Error searching space:", err)
			os.Exit(1)
		}
		defer resp.Body.Close()

		body, _ := ioutil.ReadAll(resp.Body)

		var searchResult SearchResult
		json.Unmarshal(body, &searchResult)

		for _, page := range searchResult.Results {
			searchPage(page.ID, client, accessToken)
		}

		if searchResult.Size < limit {
			break
		}

		start += limit
	}
}

func searchPage(pageID string, client *http.Client, accessToken string) {
	url := fmt.Sprintf("%s/rest/api/content/%s?expand=body.storage", confluenceURL, pageID)
	req, _ := http.NewRequest("GET", url, nil)
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", accessToken))

	resp, err := client.Do(req)
	if err != nil {
		fmt.Println("Error searching page:", err)
		os.Exit(1)
	}
	defer resp.Body.Close()

	body, _ := ioutil.ReadAll(resp.Body)

	var page Page
	json.Unmarshal(body, &page)

	content := page.Body.Storage.Value
	title := page.Title

	foundSecrets := []string{}

	for _, pattern := range patterns {
		re := regexp.MustCompile(pattern)
		matches := re.FindAllString(content, -1)

		if len(matches) > 0 {
			foundSecrets = append(foundSecrets, matches...)
		}
	}

	if len(foundSecrets) > 0 {
		fmt.Printf("Secrets found in \"%s\" (Page ID: %s):\n", title, pageID)
		for _, secret := range foundSecrets {
			fmt.Printf("  - %s\n", secret)
		}
	}
}

File: python/confluence_secrets_scanner.py
import os
import re
import requests
from requests.auth import HTTPBasicAuth

confluence_url = 'https://your-confluence-instance.com'
username = os.getenv('CONFLUENCE_USERNAME')
api_token = os.getenv('CONFLUENCE_API_TOKEN')

if username is None or api_token is None:
    print("Error: Please set the CONFLUENCE_USERNAME and CONFLUENCE_API_TOKEN environment variables.")
    exit(1)

auth = HTTPBasicAuth(username, api_token)
headers = {'Content-Type': 'application/json'}

patterns = [
    r'[a-zA-Z0-9_\-]{32}',
    r'[a-zA-Z0-9_\-]{40}',
    r'[a-zA-Z0-9_\-]{64}',
]


def get_all_spaces():
    url = f'{confluence_url}/rest/api/space'
    response = requests.get(url, headers=headers)

    if response.status_code != 200:
        print(f"Error: Unable to fetch spaces (Status code: {response.status_code})")
        exit(1)

    try:
        spaces_data = response.json()
    except requests.exceptions.JSONDecodeError:
        print("Error: Invalid JSON response when fetching spaces.")
        exit(1)

    return [space['key'] for space in spaces_data['results']]


def get_all_pages(space_key):
    url = f'{confluence_url}/rest/api/content/search'
    query = f'space={space_key} and type=page'
    params = {'cql': query}
    response = requests.get(url, headers=headers, params=params)

    if response.status_code != 200:
        print(f"Error: Unable to fetch pages for space {space_key} (Status code: {response.status_code})")
        return []

    try:
        pages_data = response.json()
    except requests.exceptions.JSONDecodeError:
        print(f"Error: Invalid JSON response when fetching pages for space {space_key}.")
        return []

    return pages_data['results']


def search_page(page):
    url = f'{confluence_url}/rest/api/content/{page["id"]}?expand=body.storage'
    response = requests.get(url, headers=headers)

    if response.status_code != 200:
        print(f"Error: Unable to fetch page {page['title']} (ID: {page['id']}) (Status code: {response.status_code})")
        return

    try:
        page_data = response.json()
    except requests.exceptions.JSONDecodeError:
        print(f"Error: Invalid JSON response when fetching page {page['title']} (ID: {page['id']}).")
        return

    content = page_data['body']['storage']['value']

    found_secrets = []

    for pattern in patterns:
        secrets = re.findall(pattern, content)
        if secrets:
            found_secrets.extend(secrets)

    if found_secrets:
        print(f"Secrets found in \"{page['title']}\" (ID: {page['id']}):")
        for secret in found_secrets:
            print(f"  - {secret}")


def main():
    space_keys = get_all_spaces()

    for space_key in space_keys:
        pages = get_all_pages(space_key)
        for page in pages:
            search_page(page)


if __name__ == "__main__":
    main()
